### PR TITLE
FeedbackState as metadata

### PR DESCRIPTION
**Pull request by andreafioraldi**
experimental PR to have generic metadata and so allow to have feedback states as meta

### CODE REVIEW

File: libafl/src/bolts/serdeany.rs [hunk 1]
```diff
@@ -605,16 +627,26 @@ macro_rules! impl_serdeany {
...
+        impl $(< $( $lt $( : $clt $(+ $dlt )* )? ),+ >)?
+            $crate::bolts::serdeany::RegistryAtStartup
+            for $struct_name $(< $( $lt ),+ >)?
+        {
+            #[cfg_attr(any(target_os = "linux", target_os = "android"), link_section = ".init_array")]
+            #[cfg_attr(target_os = "freebsd", link_section = ".init_array")]
+            #[cfg_attr(target_os = "netbsd", link_section = ".init_array")]
+            #[cfg_attr(target_os = "openbsd", link_section = ".init_array")]
+            #[cfg_attr(target_os = "illumos", link_section = ".init_array")]
+            #[cfg_attr(any(target_os = "macos", target_os = "ios"), link_section = "__DATA,__mod_init_func")]
```
Comment (domenukk) [inline libafl/src/bolts/serdeany.rs:645]: Use `target_vendor = "apple"` instead

File: libafl/src/bolts/serdeany.rs [hunk 2]
```diff
@@ -590,12 +602,22 @@ macro_rules! create_serde_registry_for_trait {
 create_serde_registry_for_trait!(serdeany_registry, crate::bolts::serdeany::SerdeAny);
 pub use serdeany_registry::*;
 
+#[cfg(feature = "std")]
+#[allow(missing_docs)]
```
Comment (domenukk) [inline libafl/src/bolts/serdeany.rs:606]: Why not document?

File: libafl/src/stages/calibrate.rs [hunk 3]
```diff
@@ -27,9 +29,10 @@ where
     I: Input,
     O: MapObserver,
     OT: ObserversTuple<I, S>,
-    S: HasCorpus<I> + HasMetadata,
+    S: HasCorpus<I> + HasMetadata + HasNamedMetadata,
 {
     map_observer_name: String,
+    map_name: String,
```
Comment (tokatoka) [inline libafl/src/stages/calibrate.rs:35]: I think here, we need map_observer_name only (?)
we want the map observer, and the feedbackstate with the same name.

Comment (domenukk) [reply libafl/src/stages/calibrate.rs:35]: I would argue there can be multiple feedbacks built on top of a single observer

File: libafl/src/feedbacks/map.rs [hunk 4]
```diff
@@ -393,11 +371,14 @@ where
         let initial = observer.initial();
 
         let map_state = state
-            .feedback_states_mut()
-            .match_name_mut::<MapFeedbackState<T>>(&self.name)
+            .named_metadata_mut()
+            .get_mut::<MapFeedbackMetadata<T>>(&self.name)
             .unwrap();
+        if map_state.history_map.len() < observer.len() {
```
Comment (tokatoka) [inline libafl/src/feedbacks/map.rs:377]: why do we resize here when map_state.history_map.len() < observer.len() ?
I think it should error out just like `assert!(size <= observer.len())`

File: libafl/src/feedbacks/map.rs [hunk 5]
```diff
@@ -393,11 +371,14 @@ where
...
+            map_state.history_map.resize(observer.len(), T::default());
+        }
 
-        assert!(size <= map_state.history_map.len(), "The size of the associated map observer cannot exceed the size of the history map of the feedback. If you are running multiple instances of slightly different fuzzers (e.g. one with ASan and another without) synchronized using LLMP please check the `configuration` field of the LLMP manager.");
+        // assert!(size <= map_state.history_map.len(), "The size of the associated map observer cannot exceed the size of the history map of the feedback. If you are running multiple instances of slightly different fuzzers (e.g. one with ASan and another without) synchronized using LLMP please check the `configuration` field of the LLMP manager.");
 
         assert!(size <= observer.len());
```
Comment (tokatoka) [inline libafl/src/feedbacks/map.rs:383]: it's unrelated to PR change.
but I think observer.len() >= size (observer.usable_count()) is always true(?)

File: libafl/src/feedbacks/map.rs (hunk 4 above)
Comment (domenukk) [reply libafl/src/feedbacks/map.rs:377]: Some map observers can potentially grow over time, the history map has to change accordingly

Comment (tokatoka) [reply libafl/src/feedbacks/map.rs:377]: ok now I understand.

File: fuzzers/tutorial/src/metadata.rs [hunk 6]
```diff
@@ -45,9 +45,13 @@ impl<S> Feedback<PacketData, S> for PacketLenFeedback
 where
     S: HasClientPerfMonitor,
 {
+    type FeedbackState = NopFeedbackState;
```
Comment (domenukk) [inline fuzzers/tutorial/src/metadata.rs:48]: Why do we still have the NopFeedbackState if we don't need a feedbackstate anymore?

Comment (domenukk) [reply fuzzers/tutorial/src/metadata.rs:48]: Or is it just not tidied up yet?

File: libafl/src/stages/calibrate.rs (hunk 3 above)
Comment (domenukk) [reply libafl/src/stages/calibrate.rs:35]: However, it may be better in general to hand in a ref to the Observer, here 🤔

**Review by domenukk — APPROVED** [review]
Looks nice!

File: libafl/src/stages/calibrate.rs (hunk 3 above)
Comment (andreafioraldi) [reply libafl/src/stages/calibrate.rs:35]: after all this time, you still have to get how the borrow checker works lol

Comment (domenukk) [reply libafl/src/stages/calibrate.rs:35]: No, why? Just pass in the observer to read the name from it, then drop the borrow again

Comment (andreafioraldi) [reply libafl/src/stages/calibrate.rs:35]: why sould you pass feedback and observer when feedback already contains this info?

File: libafl/src/feedbacks/map.rs (hunk 5 above)
Comment (andreafioraldi) [reply libafl/src/feedbacks/map.rs:383]: only for stdmapobserver

File: libafl/src/stages/calibrate.rs (hunk 3 above)
Comment (andreafioraldi) [reply libafl/src/stages/calibrate.rs:35]: or you should pass observer + feedback name as string

Comment (domenukk) [reply libafl/src/stages/calibrate.rs:35]: To create feedback, as we already do.
Nevermind didn't mean to confuse anyone 😅 it's good as it is

**Pull request merged**